truffle/src/main/ruby/core/rubinius/common/array.rb

```diff
@@ -612,6 +612,29 @@ def flatten!(level=-1)
     nil
   end
 
+
+  def inspect
+    return "[]".force_encoding(Encoding::US_ASCII) if @total == 0
+    comma = ", "
+    result = "["
+
+    return "[...]" if Thread.detect_recursion self do
+      each_with_index do |element, index|
+        temp = element.inspect
+        result.force_encoding(temp.encoding) if index == 0
+        result << temp << comma
+      end
+    end
+
+    Rubinius::Type.infect(result, self)
+    #result.shorten!(2) # MODIFIED We don't implement string shorten.
```

Comment:
@chrisseaton Yes, I made this change. I'm adding `MODIFIED` in the comments if I change things.

It looks like String#shorten! is only used in 2 places: io and here.  So, I'm a little hesitant to add it to the string public methods.

Alternatively, I could leave this as byteslice or rewrite this method a little. I think the concatenating part of this method is a little ugly. I would write it something like this:
```ruby
  def inspect
    return "[]".force_encoding(Encoding::US_ASCII) if @total == 0
    elements = Array.new(self.size + 2)    
    elements[0] = "["
    elements[self.size - 1] = "]"

    first_encoding = nil
    return "[...]" if Thread.detect_recursion self do
      each_with_index do |element, index|
        temp = element.inspect
        first_encoding = temp.encoding if index == 0
        elements[i] = temp
      end
    end
 
    result = elements.join(", ").force_encoding(first_encoding)
    Rubinius::Type.infect(result, self)
    result
  end
```
I'll do whatever you prefer: adding shorten!, leave as is, or rewriting.